spring-integration-core/src/main/java/org/springframework/integration/splitter/AbstractMessageSplitter.java

```diff
@@ -145,4 +172,81 @@ public String getComponentType() {
 	 */
 	protected abstract Object splitMessage(Message<?> message);
 
+	/**
+	 * A {@link HashMap} implementation that allows adding headers but disallows removal.
+	 *
+	 * @since 4.2
+	 *
+	 */
+	@SuppressWarnings("serial")
+	protected class PutOnlyHeaders extends HashMap<String, Object> {
+
+		public PutOnlyHeaders(Map<? extends String, ? extends Object> m) {
+			super(m);
```

Comment:
`MessageHeaders` is immutable so we have to copy headers anyway.

I added this to try to assuage some of your concerns :smile: ; I'm ok with a simple `HashMap`.